linkchecker.go

```diff
@@ -0,0 +1,437 @@
+
+	return !strings.HasPrefix(strings.ToLower(link), "mailto:") && !strings.HasPrefix(strings.ToLower(link), "ftp:")
+}
+
+func (l *LinkChecker) IsHeaderAvailable(site string) (bool, error) {
+
+	resp, err := l.HTTPClient.Head(site)
+	if err != nil {
+		return false, err
+	}
+
+	defer resp.Body.Close()
+
+	return true, nil
+}
+
+type CheckLink struct {
+	Locker sync.Mutex
+	List   map[string]bool
+}
+
+func (l *LinkChecker) IsCrawled(site string) bool {
+
+	result := false
+	if l.CheckLink.List[site] {
+		result = true
+	}
+	return result
+}
+
+func (l *LinkChecker) AddSite(site string) {
+
+	l.CheckLink.Locker.Lock()
+	defer l.CheckLink.Locker.Unlock()
+	l.CheckLink.List[site] = true
+}
+
+func (l *LinkChecker) CanonicaliseUrl(site string) (string, error) {
+
+	isUrl := false
+	newUrl := site
+	var err error
+
+	// if no initial scheme from Check, try https and http
+	if l.Scheme == "" {
+
+		schemes := []string{"https", "http"}
+
+		for _, scheme := range schemes {
+			str := []string{scheme, "://", site}
+			newUrl = strings.Join(str, "")
+			isUrl, err = l.IsHeaderAvailable(newUrl)
+			if err != nil {
+				fmt.Fprintf(l.errorLog, "unable to use https scheme for %s, %s", site, err)
+			}
+
+			if isUrl {
+				l.Scheme = scheme
+				l.Domain = site
+				break
+			}
+
+		}
+
+	}
+
+	return newUrl, nil
+}
+
+func (l *LinkChecker) CanonicaliseChildUrl(site string) (string, error) {
+
+	newUrl := site
+	var err error
+
+	url, err := url.Parse(site)
+
+	if err != nil {
+		return "", err
+	}
+
+	scheme := ""
+	if url.Scheme == "" {
+		scheme = l.Scheme + "://"
+	}
+
+	domain := ""
+	if url.Host == "" {
+		domain = l.Domain + "/"
+		//slash = "/"
+	}
+
+	// handle ./ link back to main page
+	if site == "./" {
+		// set to empty string and just use domain name
+		site = ""
+		//slash = ""
+	} else if strings.Index(site, "/") == 0 {
+		site = RemoveLeadingSlash(site)
+	}
+
+	// //slash needed for intrasite links with w/o domain (e.g <a href="home">)
+	// slash := ""
+	// assemble url
+	str := []string{scheme, domain, site}
+	newUrl = strings.Join(str, "")
+
+	return newUrl, nil
+}
+
+func RemoveLeadingSlash(site string) string {
+
+	for strings.Index(site, "/") == 0 {
+		site = strings.TrimPrefix(site, "/")
+	}
+
+	return site
+}
+
+// CLI
+func RunCLI() {
+
+	if len(os.Args) < 2 {
+		help(os.Args[0])
+		os.Exit(1)
+	}
+
+	site := os.Args[1]
+	if site == "help" {
+		fmt.Println(os.Args[0])
+		help(os.Args[0])
+		os.Exit(0)
+	}
+
+	l, err := NewLinkChecker()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+	}
+	done := make(chan bool)
+
+	go func() {
+		for {
+			r, more := <-l.Results
+			if more {
+				fmt.Println(r)
+			} else {
+				fmt.Println("received all results")
+				done <- true
+				return
+			}
+		}
+
+	}()
+
+	err = l.Check(site)
+	if err != nil {
+		fmt.Fprintln(l.errorLog, err)
+	}
+
+	<-done
+
+}
+
+func help(cliArg string) {
+
+	arg := "./linkchecker"
+
+	// bit of a hack to handle when calling from go run cmd/main.go
```

Comment:
Obviously a comment like this cannot stand 😆 

What are we really doing here? A CLI usage message is essentially for when the binary has been run with incorrect arguments. Showing `os.Args[0]` as the name of the program is just fine, whatever that happens to be. If the program was run with `go run`, it will be some path to a temporary binary, and that's okay.